mistralrs-bench/src/main.rs

```diff
@@ -0,0 +1,320 @@
+                .map(|u| u.avg_prompt_tok_per_sec)
+                .sum::<f32>();
+            sum_of_avg / result.usages.len() as f32
+        }
+        TestName::Gen(_) => {
+            // let tokens = result
+            //     .usages
+            //     .iter()
+            //     .map(|u| u.completion_tokens)
+            //     .sum::<usize>();
+            // let time = result
+            //     .usages
+            //     .iter()
+            //     .map(|u| u.total_completion_time_sec)
+            //     .sum::<f32>();
+
+            // tokens as f32 / time
+
+            let sum_of_avg = result
+                .usages
+                .iter()
+                .map(|u| u.avg_compl_tok_per_sec)
+                .sum::<f32>();
+
+            sum_of_avg / result.usages.len() as f32
+        }
+    }
+}
+
+fn print_usage(model: &str, device: &Device, results: Vec<BenchResult>) {
+    let backend = match device {
+        Device::Cpu => "CPU",
+        Device::Cuda(_) => "CUDA",
+        Device::Metal(_) => "Metal",
+    };
+    let results: Vec<Vec<CellStruct>> = results
+        .into_iter()
+        .map(|r| {
+            vec![
+                model.cell(),
+                backend.cell(),
+                r.test_name.to_string().cell(),
+                get_tok_s(&r).cell().justify(Justify::Right),
+                r.concurrency.cell().justify(Justify::Right),
+                (get_tok_s(&r) * r.concurrency as f32)
+                    .cell()
+                    .justify(Justify::Right),
+            ]
+        })
+        .collect();
+
+    let table = results
+        .table()
+        .title(vec![
+            "model".cell().bold(true),
+            // "size".cell().bold(true),
+            // "params".cell().bold(true),
+            "backend".cell().bold(true),
+            // "ngl".cell().bold(true),
+            "test".cell().bold(true),
+            "t/s".cell().bold(true),
+            "concurrency".cell().bold(true),
+            "tp/s".cell().bold(true),
+        ])
+        .bold(true);
+    print_stdout(table).expect("print table");
+}
+
+#[derive(Parser)]
+#[command(version, about, long_about = None)]
+struct Args {
+    /// Number of prompt tokens to run.
+    #[arg(long, short = 'p', default_value_t = 512)]
+    n_prompt: usize,
+
+    /// Number of generations tokens to run.
+    #[arg(long, short = 'n', default_value_t = 128)]
+    n_gen: usize,
+
+    /// Number of concurrent requests to run.
+    #[arg(long, short, default_value_t = 1)]
+    concurrency: usize,
+
+    /// Number of times to repeat each test.
+    #[arg(long, short, default_value_t = 5)]
+    repetitions: usize,
+}
+
+fn main() -> anyhow::Result<()> {
+    let args = Args::parse();
+    let tok_model_id = "mistralai/Mistral-7B-Instruct-v0.1";
+    let tokenizer_json = None;
+    let quantized_model_id = "TheBloke/Mistral-7B-Instruct-v0.1-GGUF";
+    let quantized_filename = "mistral-7b-instruct-v0.1.Q4_K_M.gguf";
+    let repeat_last_n = 64;
```

Comment:
Model was hardcoded, I don't know how to re-use the `model_selected.rs` file here.

IDK if it's best to create a new library, move it to core, make server also a lib...

Reply:
I think it would be best here to re-use the `model_selected.rs` file. Unfortunately, the logic for the server is, as you pointed out, duplicated here. I think this is something we can avoid: perhaps by adding a method to the `MistralRs` struct which acts like a constructor and takes the `ModelSelected`?